src/v/cluster/client_quota_store.cc

```diff
@@ -0,0 +1,50 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "client_quota_store.h"
+
+namespace cluster::client_quota {
+
+void store::set_quota(
+  const entity_key& key, const std::optional<entity_value>& value) {
+    if (value && !value->is_empty()) {
+        _quotas[key] = value.value();
+    } else {
+        _quotas.erase(key);
+    }
```

Comment:
thought: There seems to be some overlap here between "empty `entity_value`" as a wholesale removal action vs. `remove_entity_value`. Might make sense to restrict `entity_value` to add/modify actions and leave various removals to `remove_entity_value`.

Reply:
This will be clearer in the next PR which adds the controller command. I'm planning to use `entity_value` for storing the current state in `quota_store` and for add/modify changes in controller commands, while `remove_entity_value` will only be used in the controller commands for removal requests. The controller backend will then turn the removal/alter commands into calls to the `quota_store::set_quota`.

Reply:
gotcha, makes sense.